rmqproducer/rabbitmq_producer.py

```diff
@@ -0,0 +1,355 @@
+import sys
+import pika
+import signal
+import logging
+
+RECONNECT_TIME = 5
+
+
+class Publisher(object):
+    """This is a safe and robust publisher that will handle unexpected interactions
+    with RabbitMQ such as channel and connection closures.
+
+    If RabbitMQ closes the connection, it will reopen it. You should
+    look at the output, as there are limited reasons why the connection may
+    be closed, which usually are tied to permission related issues or
+    socket timeouts.
+
+    It uses delivery confirmations and keeps track of messages that have been 
+    sent and if they've been confirmed by RabbitMQ.
+
+    """
+
+    def __init__(self, amqp_url):
+        """Create a new instance of the Publisher class, passing in the
+        parameters used to connect to RabbitMQ.
+
+        :param str amqp_url: The AMQP url to connect with
+
+        """
+        self._connection = None
+        self._channel = None
+        self._messages = {}
+        self._message_number = 0
+        self._closing = False
+        self._LOGGER = logging.getLogger(__name__)
+        self._url = amqp_url
+
+    def parse_input_args(self, kwargs):
+        """Parse and set connection parameters from a dictionary.
+
+        Assigns defaults for missing parameters.
+        """
+        self.exchange = kwargs.get('exchange', 'default_exchange')
+        self.exchange_type = kwargs.get('exchange_type', 'topic')
+        self.exchange_durable = kwargs.get('exchange_durable', True)
+        self.exchange_auto_delete = kwargs.get('exchange_auto_delete', False)
+        self.exchange_internal = kwargs.get('exchange_internal', False)
+        self.delivery_confirmation = kwargs.get('delivery_confirmation', True)
+        self.nack_callback = kwargs.get('nack_callback')
+        self.safe_stop = kwargs.get('safe_stop', True)
+
+    def connect(self):
+        """This method connects to RabbitMQ, returning the connection handle.
+        When the connection is established, the on_connection_open method
+        will be invoked by pika. 
+
+        Since we want the reconnection to work, we have set stop_ioloop_on_close 
+        to False, which is not the default behavior of this adapter.
+
+        :rtype: pika.SelectConnection
+
+        """
+        self._LOGGER.info('Connecting to %s', self._url)
+        self._connection = pika.SelectConnection(pika.URLParameters(self._url),
```

Comment:
Select connection also provides an optional argument on_open_error_callback which will be called when the connection fails to open. Shouldn't we be defining that? We can attempt to retry connection and also log in that callback. I'm not a 100% sure about this, but [this](http://comments.gmane.org/gmane.comp.networking.rabbitmq.general/46693) thread gives some details.